### PR TITLE
[jit] Add support for `del`

### LINKED ISSUE #20615: [jit] del statements aren't supported
`del` isn't currently supported for dicts or lists:

```python
@torch.jit.script
def fn(x):
    # type: (Dict[str, int])
    x.pop("hi")		# ok
    del x["hello"]	# torch.jit.frontend.UnsupportedNodeError: del statements aren't supported
```

`del` can also be used on lists:
```python
@torch.jit.script
def fn(x):
    # type: (List[str])
    del x[0]		# torch.jit.frontend.UnsupportedNodeError: del statements aren't supported:
```

cc @suo

### PR DESCRIPTION
**Pull request by driazati**
Adds the `del` keyword to the parser and corresponding `aten::Delete` op for lists and dicts

Fixes #20615

Differential Revision: [D19181473](https://our.internmc.facebook.com/intern/diff/19181473/)

### CODE REVIEW

File: torch/csrc/jit/register_prim_ops.cpp [hunk 1]
```diff
@@ -1662,6 +1662,18 @@ int listClear(Stack& stack) {
   return 0;
 }
 
+template <typename T>
+int listDelete(Stack& stack) {
+  auto idx = pop(stack).toInt();
+  c10::List<T> list = pop(stack).to<c10::List<T>>();
+  if (idx > int64_t(list.size()) - 1) {
+    AT_ERROR("list assignment index out of range");
```
Comment (albanD) [inline torch/csrc/jit/register_prim_ops.cpp]: nit: list deletion ?

Comment (driazati) [reply torch/csrc/jit/register_prim_ops.cpp]: That does make sense but this is the same error Python puts out

Comment (albanD) [reply torch/csrc/jit/register_prim_ops.cpp]: Ho interesting. Ok then.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -2231,6 +2243,19 @@ int dictSetDefault(Stack& stack) {
   return 0;
 }
 
+int dictDelete(Stack& stack) {
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: Can you factor this out into a common method that is reused with `dictDelete` ?

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -890,6 +890,18 @@ struct to_ir {
     graph->insertNode(continue_node);
   }
 
+  void emitDelete(const Delete& stmt) {
+    Subscript subscript(stmt.expr());
```
Comment (eellison) [inline torch/csrc/jit/script/compiler.cpp]: Expecting a subscript expr without checking is going to give an incomprehensible error if someone tries to use `del` on a variable, which is valid python.
```
a = 5
del a
```
This also doesn't handle other valid `del` cases:
```
li = [1, 2, 3]
del li[1:3]
print(a) # [1]
```

File: torch/csrc/jit/register_prim_ops.cpp [hunk 4]
```diff
@@ -1662,6 +1662,18 @@ int listClear(Stack& stack) {
   return 0;
 }
 
+template <typename T>
+int listDelete(Stack& stack) {
+  auto idx = pop(stack).toInt();
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: Same comment, factor list with `listPop`?

**Review by eellison — COMMENTED** [review]
Looks good, i don't know if we should add it without fuller support for `del` cases. Also, in this current implementation, it doesn't add any functionality to TorchScript, since we already support pop.

**Comment by driazati** [comment]
These `del` implementations are specialized anyways so even if we supported `del` for plain variables we'd need this to support it for containers. I think it's useful in that it's one more thing users won't have to change in their code if they're already using it

File: test/jit/test_list_dict.py [hunk 5]
```diff
@@ -97,6 +98,37 @@ def reassign_nested():
...
+        # it manually
+        cu = torch.jit.CompilationUnit()
+        cu.define(dedent(inspect.getsource(fn)))
+        self.assertEqual(cu.fn(inputs()), python_out)
+        self.assertEqual(torch.jit.script(fn)(inputs()), python_out)
+
+        @torch.jit.script
+        def fn2(x):
+            # type: (List[int]) -> List[int]
+            del x[100]
```
Comment (eellison) [inline test/jit/test_list_dict.py]: Can you add an error case for 
`del x[1:4]` and `x[1, 2]`

**Review by eellison — APPROVED** [review]
LGTM. Can you open an issue for supporting list deletion with slices

**Pull request closed without merging**